Packages/com.unity.inputsystem/Tests/InputSystem/Plugins/XRTests.cs

```diff
@@ -1,29 +1,648 @@
+        Assert.That(generatedLayout.extendsLayout, Is.EqualTo("XRController"));
+    }
+
+    [Test]
+    [Category("Devices")]
+    public void Devices_LeftAndRightRoleCreatesXRControllerDevice()
+    {
+        testRuntime.ReportNewInputDevice(
+            new InputDeviceDescription
+        {
+            interfaceName = XRUtilities.kXRInterface,
+            product = "XRLeftHandedDevice",
+            manufacturer = "XRManufacturer",
+            capabilities = new XRDeviceDescriptor
+            {
+                deviceRole = DeviceRole.LeftHanded,
+                inputFeatures = new List<XRFeatureDescriptor>()
+                {
+                    new XRFeatureDescriptor()
+                    {
+                        name = "SimpleFeature",
+                        featureType = FeatureType.Binary
+                    }
+                }
+            }.ToJson()
+        }.ToJson());
+
+        testRuntime.ReportNewInputDevice(
+            new InputDeviceDescription
+        {
+            interfaceName = XRUtilities.kXRInterface,
+            product = "XRRightHandedDevice",
+            manufacturer = "XRManufacturer",
+            capabilities = new XRDeviceDescriptor
+            {
+                deviceRole = DeviceRole.RightHanded,
+                inputFeatures = new List<XRFeatureDescriptor>()
+                {
+                    new XRFeatureDescriptor()
+                    {
+                        name = "SimpleFeature",
+                        featureType = FeatureType.Binary
+                    }
+                }
+            }.ToJson()
+        }.ToJson());
+
+        InputSystem.Update();
+
+        Assert.That(InputSystem.devices, Has.Count.EqualTo(2));
+
+        var leftHandedDevice = InputSystem.devices[0];
+        Assert.That(leftHandedDevice, Is.TypeOf<XRController>());
+        Assert.That(leftHandedDevice.description.interfaceName, Is.EqualTo(XRUtilities.kXRInterface));
+        Assert.That(leftHandedDevice.usages.Count, Is.EqualTo(1));
+        Assert.That(leftHandedDevice.usages[0], Is.EqualTo(CommonUsages.LeftHand));
+        Assert.That(XRController.leftHand, Is.EqualTo(leftHandedDevice));
+
+        var rightHandedDevice = InputSystem.devices[1];
+        Assert.That(rightHandedDevice, Is.TypeOf<XRController>());
+        Assert.That(rightHandedDevice.description.interfaceName, Is.EqualTo(XRUtilities.kXRInterface));
+        Assert.That(rightHandedDevice.usages.Count, Is.EqualTo(1));
+        Assert.That(rightHandedDevice.usages[0], Is.EqualTo(CommonUsages.RightHand));
+        Assert.That(XRController.rightHand, Is.EqualTo(rightHandedDevice));
+    }
+
+    [Test]
+    [Category("Devices")]
+    public void Devices_HardwareTrackerCreatesDefaultInputDevice()
+    {
+        testRuntime.ReportNewInputDevice(
+            new InputDeviceDescription
+        {
+            interfaceName = XRUtilities.kXRInterface,
+            product = "HardwareTracker",
```

Comment:
What's a "HardwareTracker"?

Reply:
A hardware tracker is the hockey puck device we have that is attached to objects to render in game.  https://www.vive.com/us/vive-tracker/